### PR TITLE
DAOS-13493 rebuild: skip invalid shard and target

### PR DESCRIPTION
**Pull request by wangdi1**
Skip invalid shard and target for upgrade migration.

remove invalid assertion for rebuild retry.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by liw — DISMISSED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Comment by wangdi1** [comment]
ping

File: src/container/srv_container.c [hunk 1]
```diff
@@ -4563,6 +4563,10 @@ upgrade_cont_cb(daos_handle_t ih, d_iov_t *key, d_iov_t *val, void *varg)
 		}
 	}
 
+	entry = daos_prop_entry_get(prop, DAOS_PROP_CO_OBJ_VERSION);
+	D_ASSERT(entry != NULL);
+	entry->dpe_val = DS_POOL_OBJ_VERSION;
+	entry->dpe_flags &= ~DAOS_PROP_ENTRY_NOT_SET;
```
Comment (liw) [inline src/container/srv_container.c:4569]: @wangshilong, could you help review this PR, please? Also, do you know if we need to do this (i.e., to update `prop`) for the other DB updates in this function?

**Review by liw — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Comment by wangdi1** [comment]
ping @daos-stack/iof-gatekeeper  thanks.

File: src/container/srv_container.c (hunk 1 above)
Comment (wangshilong) [reply src/container/srv_container.c:4569]: I don't think we need do this, this is only needed when server returning prop to clients, so client know even this prop has value, but it is missed.

Comment (liuxuezhao) [reply src/container/srv_container.c:4569]: @wangshilong I was confused what's the purpose of those code either.
But it actually changed the prop's OBJ_VERSION entry and in the below code -
4703 >       >       >       rc = cont_iv_prop_update(ap->svc->cs_pool->sp_iv_ns,                        
4704 >       >       >       >       >       >        cont_uuid, prop, true);   
it update the prop to IV prop cache, seems that is necessary? (keep IV prop cache consistent with local RDB).

Comment (liw) [reply src/container/srv_container.c:4569]: Yep, if other DB updates in this function do not need this, then I'm also confused by why OBJ_VERSION needs this.

Comment (wangshilong) [reply src/container/srv_container.c:4569]: > @wangshilong I was confused what's the purpose of those code either. But it actually changed the prop's OBJ_VERSION entry and in the below code - 4703 > > > rc = cont_iv_prop_update(ap->svc->cs_pool->sp_iv_ns, 4704 > > > > > > cont_uuid, prop, true); it update the prop to IV prop cache, seems that is necessary? (keep IV prop cache consistent with local RDB).

You are right, we need this.

Comment (wangdi1) [reply src/container/srv_container.c:4569]: Yes, the entry is used to update IV cache.

>if other DB updates in this function do not need this, then I'm also confused by why OBJ_VERSION needs this.

This is upgrade callback. so global pool version and object layout version are supposed to be updated. both for DB and IV cache.

**Review by gnailzenh — APPROVED** [review]

**Pull request merged**